### PR TITLE
Reduce the size of HeaderDescriptor

### LINKED ISSUE #62847: Consider reducing HeaderDescriptor size by using object field
Currently, the HeaderDescriptor struct has two fields:

```c#
        private readonly string _headerName;
        private readonly KnownHeader? _knownHeader;
```

A HeaderDescriptor is either:
(1) a known header, in which case KnownHeader is not null and _headerName is equal to _knownHeader.Name; or
(2) a custom header, in which case KnownHeader is null

We use HeaderDescriptor as the key for the HttpHeader header dictionary, which means we end up having a lot of instances of this structure. 

We could optimize for memory here by combining these two fields into a single object field that contains either a string or a KnownHeader, and then do type casts. This would result in a small amount of additional CPU work to do the casts, but this may be worth the memory savings.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #62847

**NonValidatedAdd_NonValidatedEnumerate**:

|   Toolchain | RequestHeaders |        Mean |     Error |     StdDev |      Median | Ratio | RatioSD |  Gen 0 | Allocated |
|------------ |--------------- |------------:|----------:|-----------:|------------:|------:|--------:|-------:|----------:|
|        main |              2 |   118.64 ns |  0.274 ns |   1.300 ns |   118.28 ns |  1.00 |    0.00 | 0.0253 |     256 B |
|   unsafe-pr |              2 |    94.69 ns |  0.764 ns |   3.680 ns |    95.17 ns |  0.80 |    0.03 | 0.0222 |     224 B |
|          pr |              2 |    92.44 ns |  0.269 ns |   1.303 ns |    92.51 ns |  0.78 |    0.01 | 0.0222 |     224 B |
|             |                |             |           |            |             |       |         |        |           |
|        main |              4 |   210.50 ns |  0.653 ns |   3.068 ns |   208.73 ns |  1.00 |    0.00 | 0.0253 |     256 B |
|   unsafe-pr |              4 |   157.14 ns |  0.234 ns |   1.118 ns |   156.71 ns |  0.75 |    0.01 | 0.0222 |     224 B |
|          pr |              4 |   165.77 ns |  0.813 ns |   3.949 ns |   165.00 ns |  0.79 |    0.03 | 0.0222 |     224 B |
|             |                |             |           |            |             |       |         |        |           |
|        main |              6 |   356.20 ns |  0.488 ns |   2.366 ns |   354.75 ns |  1.00 |    0.00 | 0.0467 |     472 B |
|   unsafe-pr |              6 |   279.82 ns |  0.794 ns |   3.753 ns |   277.51 ns |  0.79 |    0.01 | 0.0372 |     376 B |
|          pr |              6 |   294.23 ns |  0.400 ns |   1.970 ns |   293.46 ns |  0.83 |    0.01 | 0.0372 |     376 B |
|             |                |             |           |            |             |       |         |        |           |
|        main |              8 |   484.50 ns |  0.629 ns |   3.003 ns |   482.65 ns |  1.00 |    0.00 | 0.0467 |     472 B |
|   unsafe-pr |              8 |   347.89 ns |  1.147 ns |   5.490 ns |   344.33 ns |  0.72 |    0.01 | 0.0372 |     376 B |
|          pr |              8 |   355.82 ns |  1.256 ns |   6.072 ns |   353.78 ns |  0.73 |    0.02 | 0.0372 |     376 B |
|             |                |             |           |            |             |       |         |        |           |
|        main |             12 |   903.92 ns |  1.075 ns |   5.127 ns |   907.11 ns |  1.00 |    0.00 | 0.0868 |     880 B |
|   unsafe-pr |             12 |   732.21 ns |  0.769 ns |   3.643 ns |   731.08 ns |  0.81 |    0.00 | 0.0648 |     656 B |
|          pr |             12 |   779.44 ns |  2.077 ns |  10.180 ns |   774.05 ns |  0.86 |    0.02 | 0.0648 |     656 B |
|             |                |             |           |            |             |       |         |        |           |
|        main |             16 | 1,358.70 ns |  0.430 ns |   2.032 ns | 1,358.55 ns |  1.00 |    0.00 | 0.0858 |     880 B |
|   unsafe-pr |             16 | 1,170.03 ns |  7.485 ns |  39.010 ns | 1,164.91 ns |  0.86 |    0.03 | 0.0648 |     656 B |
|          pr |             16 | 1,259.05 ns |  5.470 ns |  28.508 ns | 1,241.47 ns |  0.93 |    0.02 | 0.0648 |     656 B |
|             |                |             |           |            |             |       |         |        |           |
|        main |             32 | 3,889.28 ns | 17.269 ns |  90.002 ns | 3,887.74 ns |  1.00 |    0.00 | 0.1640 |   1,672 B |
|   unsafe-pr |             32 | 3,580.34 ns | 13.858 ns |  72.100 ns | 3,570.95 ns |  0.92 |    0.03 | 0.1183 |   1,192 B |
|          pr |             32 | 4,046.49 ns | 23.530 ns | 122.632 ns | 4,030.22 ns |  1.04 |    0.03 | 0.1183 |   1,192 B |

**SendAsync (Create and serialize the request, parse the response and enumerate 4 response headers)**:

|   Toolchain | RequestHeaders |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |  Gen 0 | Allocated |
|------------ |--------------- |---------:|----------:|----------:|---------:|------:|--------:|-------:|----------:|
|        main |              4 | 1.644 μs | 0.0062 μs | 0.0311 μs | 1.630 μs |  1.00 |    0.00 | 0.0935 |     944 B |
| original-pr |              4 | 1.639 μs | 0.0057 μs | 0.0281 μs | 1.630 μs |  1.00 |    0.03 | 0.0858 |     880 B |
|          pr |              4 | 1.650 μs | 0.0038 μs | 0.0191 μs | 1.653 μs |  1.00 |    0.02 | 0.0858 |     880 B |
|             |                |          |           |           |          |       |         |        |           |
|        main |              8 | 2.084 μs | 0.0013 μs | 0.0064 μs | 2.082 μs |  1.00 |    0.00 | 0.1144 |   1,160 B |
| original-pr |              8 | 2.055 μs | 0.0030 μs | 0.0145 μs | 2.055 μs |  0.99 |    0.01 | 0.0992 |   1,032 B |
|          pr |              8 | 2.026 μs | 0.0068 μs | 0.0327 μs | 2.028 μs |  0.97 |    0.02 | 0.0992 |   1,032 B |
|             |                |          |           |           |          |       |         |        |           |
|        main |             16 | 3.428 μs | 0.0039 μs | 0.0185 μs | 3.424 μs |  1.00 |    0.00 | 0.1526 |   1,568 B |
| original-pr |             16 | 3.328 μs | 0.0044 μs | 0.0209 μs | 3.324 μs |  0.97 |    0.00 | 0.1297 |   1,312 B |
|          pr |             16 | 3.419 μs | 0.0141 μs | 0.0675 μs | 3.446 μs |  1.00 |    0.02 | 0.1297 |   1,312 B |

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/QPackStaticTable.cs [hunk 1]
```diff
@@ -96,7 +96,7 @@ internal static class QPackStaticTable
             (new HeaderDescriptor(KnownHeaders.Authorization), ""), // 84
             (new HeaderDescriptor(KnownHeaders.ContentSecurityPolicy), "script-src 'none'; object-src 'none'; base-uri 'none'"), // 85
             (new HeaderDescriptor("early-data"), "1"), // 86
-            (new HeaderDescriptor("expect-ct"), ""), // 87
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/QPackStaticTable.cs:99]: This is an existing bug in HTTP/3 where `x-xss-protection` and `expect-ct` lookups/removes would silently fail if added via `OnStaticIndexedHeader`.

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs [hunk 2]
```diff
@@ -15,32 +16,71 @@ namespace System.Net.Http.Headers
...
+                    Unsafe.As<string>(descriptor);
+            }
+        }
+
+        public bool IsKnownHeader([NotNullWhen(true)] out KnownHeader? knownHeader, [NotNullWhen(false)] out string? headerName)
+        {
+            object descriptor = _descriptor;
+            if (descriptor.GetType() == typeof(KnownHeader))
+            {
+                knownHeader = Unsafe.As<KnownHeader>(descriptor);
```
Comment (danmoseley) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:52]: How is this different to `knownHeader = (Knownheader)descriptor;` ?

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs [hunk 3]
```diff
@@ -15,32 +16,71 @@ namespace System.Net.Http.Headers
...
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public string Name
+        {
+            get
+            {
+                object descriptor = _descriptor;
+                return descriptor.GetType() == typeof(KnownHeader) ?
+                    Unsafe.As<KnownHeader>(descriptor).Name :
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:42]: KnownHeader is sealed; you should get almost the same codegen with an `is` cast and if for some reason you don't, that's something that should be fixable in the JIT. The main difference I'd expect `is` will still incur a null check, but that also means you get the added protection of not getting a null reference exception if default(HeaderDescriptor).Name is used. And with all the headers that are "known", I'd hope the impact of the cast to a sealed string (`descriptor as string`) for the ones that aren't should not be impactful. All this unsafe casting isn't ideal.

What percentage of the improvements you showed are due to the use of Unsafe.As? e.g. if you instead use patterns like:
```C#
public string Name
{
    get
    {
        object descriptor = _descriptor;
        return descriptor is KnownHeader header ? header.Name : descriptor as string;
    }
}
```
what does that do to the numbers?

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:42]: The [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlrFiKbZB5WNgAmEJoICe5XBigMwGcgAkY+kQBFZYKKz4ZoygN7LyAeR87hLYGDA6DgZGphDAAFYw1uQA+nqu7p7QikqB5P6BxADMlFQUAHLYHAUBfrl5gQDmMBg1DXUNneRxicnpuG4eXlDkALyp/YNZUDldDcQA7OSTmcM0AOItACrGfDAAFACUY+MYuzAQAGb7ANJMEADuTPaOUMcA/G1zeQCqTLjYS6cACCuAAPHdHs8oiIAHz7FZDaCHGiVDjkEBfb4BP4AoE0UFg6ikeGI6aHWZdAC+bRpKnqAWKpQqVRgACY2h0us1Wgz2ljOosJhkkSNWLhyJCni90iMABYwkbvcgK16o1kY4UDVbQch4ZmUzp0vLGwJtJnACAQbQASVwUuhr32EAYNgdMpE5AA1vdpYq0N1XcyVYq0TBDpyBd0EkkbGThmMtVNhoaGqxLuQESLphttucjidyGc9ldbr7HbLDhG+XkudifVCPSNxrjASDwe7FaTs8MKVHfv82zRRF6PDamKwMM6g6rZWG+zWukKLAwYKmTVGYIJcDAo3Xvq38SOxxOpy6bA2/a8F9iArORGHE4f20Syt3taKb7ehZctDv14EpoBEB+R8haVq2va5ZNmy05utBireghrwBuewb3lA86RouAQ9LGyw9rq4xpIRMxRummbxrq4qSshsohtee79oEl4Vp64wYQBnTPsOo58OOk5wQxc6sl+2LLpYa5RiBeRbjuTE4Q0PHHvxp5CaxTZid8GGPuMVEjPqxJcfMSy/tuUmKSBdJ0qo6hQJo2g7loMB6JQbK0Y2iq+OaJTEuQj4+OQPJyOQ1lAA==) is quite a bit worse, but I'll remeasure without `Unsafe.As`.


> that also means you get the added protection of not getting a null reference exception if default(HeaderDescriptor).Name is used

Arguably hitting an exception here is even desired as it indicates our logic is broken / there was concurrent access to the collection that messed things up.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:42]: > The codegen is quite a bit worse

Not really... and it's way safer and easier to reason about. We should only be playing Unsafe.As tricks when it's absolutely critical. 

As an aside, you should still only read _descriptor once into a local. 

> Arguably hitting an exception here is even desired

Hitting a null reference exception in conjunction with use of Unsafe in situations where concurrent access is possible (even if not supported) is never desirable.

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs [hunk 4]
```diff
@@ -15,32 +16,71 @@ namespace System.Net.Http.Headers
...
+                Unsafe.SkipInit(out knownHeader);
+                headerName = Unsafe.As<string>(descriptor);
+                return false;
+            }
+        }
+
+        public HttpHeaderParser? Parser => (_descriptor as KnownHeader)?.Parser;
+        public HttpHeaderType HeaderType => _descriptor is KnownHeader knownHeader ? knownHeader.HeaderType : HttpHeaderType.Custom;
+
+        public bool Equals(KnownHeader other) => ReferenceEquals(_descriptor, other);
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:67]: It's a little weird to call this method "Equals".

I'd suggest calling it "IsKnownHeader(KnownHeader header)" but we already have a method called that.

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs (hunk 3 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:42]: Additionally, I would hope that this (and other routines here) are typically getting inlined. That will impact codegen in nontrivial ways.

Making these routines more complicated may actually inhibit inlining, which is very likely counterproductive.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 5]
```diff
@@ -282,7 +282,7 @@ private async ValueTask WriteHeadersAsync(HttpHeaders headers, string? cookiesFr
 
                         await WriteStringAsync(_headerValues[0], async, valueEncoding).ConfigureAwait(false);
 
-                        if (cookiesFromContainer != null && header.Key.KnownHeader == KnownHeaders.Cookie)
+                        if (cookiesFromContainer != null && header.Key.Equals(KnownHeaders.Cookie))
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:285]: We have the knownHeader from above, why not just do "knownHeader == KnownHeaders.Cookie"?

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs (hunk 3 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:42]: With the following change as `pr`: https://github.com/MihaZupan/runtime-1/commit/f67226ad1acda3d02d79f69953a143c5a06da10a

**NonValidatedAdd_NonValidatedEnumerate**

| Toolchain | RequestHeaders |        Mean |     Error |     StdDev |      Median | Ratio | RatioSD |
|---------- |--------------- |------------:|----------:|-----------:|------------:|------:|--------:|
|        pr |              2 |    91.51 ns |  0.139 ns |   0.662 ns |    91.16 ns |  1.00 |    0.00 |
|    unsafe |              2 |    89.91 ns |  0.301 ns |   1.436 ns |    88.97 ns |  0.98 |    0.02 |
|           |                |             |           |            |             |       |         |
|        pr |              4 |   158.69 ns |  0.145 ns |   0.687 ns |   158.38 ns |  1.00 |    0.00 |
|    unsafe |              4 |   160.77 ns |  0.777 ns |   4.049 ns |   161.48 ns |  1.01 |    0.02 |
|           |                |             |           |            |             |       |         |
|        pr |              6 |   296.10 ns |  1.190 ns |   6.203 ns |   292.65 ns |  1.00 |    0.00 |
|    unsafe |              6 |   274.07 ns |  0.972 ns |   4.614 ns |   275.46 ns |  0.93 |    0.02 |
|           |                |             |           |            |             |       |         |
|        pr |              8 |   354.51 ns |  1.504 ns |   7.069 ns |   350.97 ns |  1.00 |    0.00 |
|    unsafe |              8 |   345.02 ns |  1.451 ns |   6.975 ns |   340.95 ns |  0.97 |    0.03 |
|           |                |             |           |            |             |       |         |
|        pr |             12 |   781.55 ns |  2.801 ns |  13.887 ns |   775.41 ns |  1.00 |    0.00 |
|    unsafe |             12 |   729.47 ns |  3.631 ns |  17.520 ns |   729.39 ns |  0.93 |    0.03 |
|           |                |             |           |            |             |       |         |
|        pr |             16 | 1,244.16 ns |  2.477 ns |  11.713 ns | 1,237.19 ns |  1.00 |    0.00 |
|    unsafe |             16 | 1,145.68 ns |  4.649 ns |  22.433 ns | 1,156.79 ns |  0.92 |    0.03 |
|           |                |             |           |            |             |       |         |
|        pr |             32 | 3,936.57 ns |  6.690 ns |  32.917 ns | 3,923.61 ns |  1.00 |    0.00 |
|    unsafe |             32 | 3,587.23 ns | 39.021 ns | 184.894 ns | 3,466.43 ns |  0.91 |    0.05 |

**SendAsync (Create and serialize the request, parse the response and enumerate 4 response headers)**

| Toolchain | RequestHeaders |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |
|---------- |--------------- |---------:|----------:|----------:|---------:|------:|--------:|
|        pr |              4 | 1.640 μs | 0.0061 μs | 0.0301 μs | 1.639 μs |  1.00 |    0.00 |
|    unsafe |              4 | 1.622 μs | 0.0049 μs | 0.0235 μs | 1.607 μs |  0.99 |    0.02 |
|           |                |          |           |           |          |       |         |
|        pr |              8 | 2.029 μs | 0.0031 μs | 0.0148 μs | 2.032 μs |  1.00 |    0.00 |
|    unsafe |              8 | 2.041 μs | 0.0051 μs | 0.0252 μs | 2.053 μs |  1.01 |    0.01 |
|           |                |          |           |           |          |       |         |
|        pr |             16 | 3.464 μs | 0.0045 μs | 0.0214 μs | 3.463 μs |  1.00 |    0.00 |
|    unsafe |             16 | 3.389 μs | 0.0072 μs | 0.0352 μs | 3.395 μs |  0.98 |    0.01 |

> Making these routines more complicated may actually inhibit inlining, which is very likely counterproductive.

At least for calls like `Equals(HeaderDescriptor)`, they're getting inlined with or without `Unsafe`.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:42]: The difference is certainly measurable when looking at header operations in isolation, but not game-changing when compared to the total cost of sending a request.

> We should only be playing Unsafe.As tricks when it's absolutely critical.

Not critical, I'm just fond of `Unsafe` :)

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs (hunk 4 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs:67]: Would you find the `==` operator more fitting here?

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs (hunk 3 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs]: Removed the Unsafe and minimized changes

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 5 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:285]: `knownHeader` is no longer in scope after 65c59000fe2e54b3f21ea510115fec2275db6e81

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs (hunk 3 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs]: I've updated the benchmarks in the initial post - the differences on `SendAsync` with/without unsafe are insignificant.

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs]: The cast is more expensive (the JIT won't skip it based on the `GetType` check above). Overall it didn't make that big of a difference so I simplified the code.

**Comment by MihaZupan** [comment]
@dotnet/ncl Please take a look

**Review by ManickaP — APPROVED** [review]
LGTM thanks

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/QPackStaticTable.cs [hunk 6]
```diff
@@ -96,7 +96,7 @@ internal static class QPackStaticTable
...
+            (new HeaderDescriptor(KnownHeaders.ExpectCT), ""), // 87
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/Headers/QPackStaticTable.cs:99]: Should we be adding KnownHeaders entries for the other `new HeaderDescriptor(string)` entries in this table?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/Headers/QPackStaticTable.cs (hunk 6 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/Headers/QPackStaticTable.cs:99]: That would be my hope as part of #63750, I updated the post to include `QPackStaticTable`

**Pull request merged**